docker2earth/convert.go

```diff
@@ -0,0 +1,116 @@
+package docker2earth
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/earthly/earthly/fileutils"
+
+	"github.com/moby/buildkit/frontend/dockerfile/instructions"
+	"github.com/moby/buildkit/frontend/dockerfile/parser"
+	"github.com/pkg/errors"
+)
+
+func getArtifactName(s string) string {
+	split := strings.Split(s, "/")
+	n := len(split)
+	return split[n-1]
+}
+
+// Docker2Earth converts an existing Dockerfile in the current directory and writes out an Earthfile in the current directory
+// and error is returned if an Earthfile already exists.
+func Docker2Earth() error {
+	if fileutils.FileExists("Earthfile") {
+		return fmt.Errorf("Earthfile already exists; please delete it if you wish to continue")
+	}
+
+	in, err := os.Open("Dockerfile")
+	if err != nil {
+		return errors.Wrap(err, "failed to open ./Dockerfile")
+	}
+	defer in.Close()
+
+	targets := [][]string{
+		{
+			"# This Earthfile was generated using docker2earth",
+			"# the conversion is done on a best-effort basis",
+			"# and might not follow best practices, please",
+			"# visit http://docs.earthly.dev for Earthfile guides",
+		},
+	}
+
+	dockerfile, err := parser.Parse(in)
+	if err != nil {
+		return errors.Wrap(err, "failed to parse Dockerfile")
+	}
+
+	stages, _, err := instructions.Parse(dockerfile.AST)
+	if err != nil {
+		return errors.Wrap(err, "failed to parse Dockerfile")
+	}
+
+	names := map[string]int{}
+
+	for i, stage := range stages {
+		targets = append(targets, []string{
+			fmt.Sprintf("FROM %s", stage.BaseName),
+		})
+		if stage.Name == "" {
+			names[fmt.Sprintf("%d", i)] = i
+		} else {
+			names[stage.Name] = i
+		}
+
+		for _, cmd := range stage.Commands {
+			l := fmt.Sprintf("%v", cmd)
+			if strings.HasPrefix(l, "COPY ") && strings.Contains(l, "--from") {
+				parts := strings.Split(l, " ")
+				if len(parts) != 4 {
+					return fmt.Errorf("failed to parse %q", l)
+				}
+				kv := strings.Split(parts[1], "=")
+				if len(kv) != 2 {
+					return fmt.Errorf("failed to parse %q", l)
+				}
+				fromStageName := kv[1]
+				n := names[fromStageName]
+				artifactName := getArtifactName(parts[2])
+				l = fmt.Sprintf("COPY +subbuild%d/%s %s", n+1, artifactName, parts[3])
+				targets[n+1] = append(targets[n+1], fmt.Sprintf("SAVE ARTIFACT %s %s\n", parts[2], artifactName))
+			}
+			if strings.HasPrefix(l, "ADD ") {
+				return fmt.Errorf("earthly does not support ADD, please convert to COPY instead")
+			}
+			targets[i+1] = append(targets[i+1], l)
+		}
+	}
+	i := len(targets) - 1
+	targets[i] = append(targets[i], "SAVE IMAGE myimage:latest")
```

Comment:
Perhaps it's better if the name for the final target is specified by the user when invoking the command and the names for any other intermediate targets are of some form representing some identifier of the stage (eg `stage-1:latest` or `stage-<stage-name>:latest`, if it has a name given by `FROM ... AS <stage-name>`).